File: wordler.py
from typing import List


LENGTH = 5
NUM_LETTERS = 26
CORRECT = 'g'
WRONG_POS = 'y'
WRONG = 'b'


class WordNode:
    def __init__(self, word: str):
        self.word = word
        self.next = None


def char_index(c: str):
    return ord(c) - 97


def read_word_list(words_file: str) -> WordNode:
    head = None
    tail = None
    with open(words_file, 'r') as f:
        for i, word in enumerate(f):
            node = WordNode(word.strip())
            if i == 0:
                head = node
                tail = node
            else:
                tail.next = node
                tail = tail.next
    return head


def valid(word: str, result_word: str, result: str) -> bool:
    for i in range(LENGTH):
        if result[i] == CORRECT:
            if word[i] != result_word[i]:
                return False
        elif result[i] == WRONG_POS:
            if word[i] == result_word[i] or result_word[i] not in word:
                return False
        else:
            if result_word[i] in word:
                return False
    return True


def filter_word_list(head: WordNode, result_word: str, result: str) -> WordNode:
    while (head is not None) and (not valid(head.word, result_word, result)):
        head = head.next
    if head is None:
        return None
    prev = head
    node = prev.next
    while node is not None:
        if valid(node.word, result_word, result):
            prev = node
        else:
            prev.next = node.next
        node = prev.next
    return head


def compute_letter_counts(head: WordNode, letter_counts: List[List[int]]):
    for i in range(LENGTH):
        for j in range(NUM_LETTERS):
            letter_counts[i][j] = 0
    while head is not None:
        for i in range(LENGTH):
            letter_counts[i][char_index(head.word[i])] += 1
        head = head.next


def find_optimal_word(head: WordNode, letter_counts: List[List[int]]) -> str:
    best_count = 0
    best_word = None
    while head is not None:
        count = 0
        for i in range(LENGTH):
            count += letter_counts[i][char_index(head.word[i])]
        if count > best_count:
            best_count = count
            best_word = head.word
        head = head.next
    return best_word


def main(words_file: str):
    word_list = read_word_list(words_file)
    letter_counts = [[0 for _ in range(NUM_LETTERS)] for _ in range(LENGTH)]
    for i in range(6):
        compute_letter_counts(word_list, letter_counts)
        best_word = find_optimal_word(word_list, letter_counts)
        print('The next word is:', best_word)
        result = input(f'What was the result? ({CORRECT} for correct, {WRONG_POS} for wrong position, {WRONG} for wrong): ')
        if result == CORRECT * LENGTH:
            break
        word_list = filter_word_list(word_list, best_word, result)


if __name__ == '__main__':
    main('five_letter_words.txt')

File: wordler.cpp
#include "wordler.h"

void clearWordList(WordNode* wordList) {
    WordNode* next;
    while (wordList != NULL) {
        next = wordList->next;
        delete wordList;
        wordList = next;
    }
}

WordNode* readWordList(std::string const & wordFile) {
    WordNode* head;
    WordNode* tail;
    bool first = true;
    std::string line;
    std::ifstream f(wordFile);
    while (std::getline(f, line)) {
        WordNode* node = new WordNode();
        node->word = line;
        if (first) {
            head = node;
            tail = node;
            first = false;
        } else {
            tail->next = node;
            tail = node;
        }
    }   
    f.close();
    return head;
}

bool valid(std::string const & word, std::string const & resultWord, std::string const & result) {
    bool found;
    for (int i = 0; i < LENGTH; i++) {
        if (result[i] == CORRECT) {
            if (word[i] != resultWord[i]) {
                return false;
            }
        } else if (result[i] == CLOSE) {
            if (word[i] == resultWord[i]) {
                return false;
            }
            found = false;
            for (int j = 0; j < LENGTH; j++) {
                if (word[j] == resultWord[i]) {
                    found = true;
                    break;
                }
            }
            if (!found) {
                return false;
            }
        } else {
            for (int j = 0; j < LENGTH; j++) {
                if (word[j] == resultWord[i]) {
                    return false;
                }
            }
        }
    }
    return true;
}

void filterWordList(WordNode* & wordList, std::string const & word, std::string const & result) {
    WordNode* prev;
    while (wordList != NULL && !valid(wordList->word, word, result)) {
        prev = wordList;
        wordList = wordList->next;
        delete prev;
    }
    if (wordList == NULL) {
        return;
    }
    prev = wordList;
    WordNode* node = prev->next;
    while (node != NULL) {
        if (valid(node->word, word, result)) {
            prev = node;
        } else {
            prev->next = node->next;
            delete node;
        }
        node = prev->next;
    }
}

void computeLetterCounts(WordNode* wordList, int letterCounts[LENGTH][NUM_LETTERS]) {
    for (int i = 0; i < LENGTH; i++) {
        for (int j = 0; j < NUM_LETTERS; j++) {
            letterCounts[i][j] = 0;
        }
    }
    while (wordList != NULL) {
        for (int i = 0; i < LENGTH; i++) {
            letterCounts[i][charIndex(wordList->word[i])]++;
        }
        wordList = wordList->next;
    }
}

std::string findOptimalWord(WordNode* wordList, int letterCounts[LENGTH][NUM_LETTERS]) {
    int bestCount = 0;
    int count;
    std::string bestWord;
    while (wordList != NULL) {
        count = 0;
        for (int i = 0; i < LENGTH; i++) {
            count += letterCounts[i][charIndex(wordList->word[i])];
        }
        if (count > bestCount) {
            bestCount = count;
            bestWord = wordList->word;
        }
        wordList = wordList->next;
    }
    return bestWord;
}

int charIndex(char c) {
    int i = c;
    return i - 97;
}

int main() {
    int letterCounts[LENGTH][NUM_LETTERS];
    WordNode* wordList = readWordList("five_letter_words.txt");
    std::string bestWord;
    std::string result;
    bool OK;
    for (int i = 0; i < 6; i++) {
        computeLetterCounts(wordList, letterCounts);
        bestWord = findOptimalWord(wordList, letterCounts);
        std::cout << "The next word to try is: " << bestWord << std::endl;
        std::cout << "What was the result? (" << CORRECT << " for correct, " << CLOSE << " for close, " << WRONG << " for wrong): ";
        std::getline(std::cin, result);
        OK = true;
        for (int j = 0; j < LENGTH; j++) {
            if (result[j] != CORRECT) {
                OK = false;
                break;
            }
        }
        if (OK) {
            break;
        }
        filterWordList(wordList, bestWord, result);
    }
    clearWordList(wordList);
 
    return 0;
}